[LISKDOCS-60] Indonesian Translation for [LISKDOCS-19] Introduction to Web3 app development (#288)
LiskHQ/lisk-docs: i18n/ind/docusaurus-plugin-content-docs/current/building-on-lisk/web3-app-development.mdx

Reviewer: @fujianggara10 Please see `docs/building-on-lisk/web3-app-development.mdx` for reference when translating. Please kindly remove line 24 - 27 as follows as it is not in the English file:
:::info[Kami menghargai kesabaran Anda]
Kami sedang berusaha menerjemahkan halaman ini. 
Mohon tunggu sementara kami menyiapkan versi **bahasa Indonesia** dari dokumen ini.
:::

---
title: Pengembangan aplikasi Web3
description: 'Pengenalan ke pengembangan aplikasi Web3 di Lisk.'
keywords:
  [
    'Lisk',
    'Aplikasi Web3',
    'Aplikasi-aplikasi web3',
    'Aplikasi Web3',
    'Pengembangan aplikasi Web3',
    'Aplikasi mobile Web3',
    'Progressive Web App Web3',
    'Pengembangan dApp',
    'Pengembangan dApp mobile',

>     'Pengembangan app mobile',

Reviewer: ```suggestion
    'Pengembangan aplikasi mobile',
```

Mohon tunggu sementara kami menyiapkan versi **bahasa Indonesia** dari dokumen ini.
:::

# Pengembangan aplikasi Web3

> Di halaman ini, Anda akan belajar tentang apa itu aplikasi Web3, bagaimana perbedaannya dari aplikasi Web 2.0, dan cara memulai pengembangan aplikasi Web3 di Lisk.

Reviewer: ```suggestion
Di halaman ini, Anda akan belajar tentang apa itu aplikasi Web3, bagaimana mereka berbeda dari aplikasi Web 2.0, dan cara memulai pengembangan aplikasi Web3 di Lisk.
```

> ## What are Web3 applications?

Reviewer: @fujianggara10 Please also retain the line spacing as in the original English doc page. There's one line spacing here before the heading. Also, this line also hasn't been translated.

```suggestion

## Apa itu aplikasi Web3?
```

### Evolusi Web

Sejak internet lahir, inovasi terus berkembang dalam cara orang berkomunikasi dan berinteraksi satu sama lain melalui internet.  
Untuk menggambarkan langkah-langkah penting dalam inovasi ini, situs web dan aplikasi umumnya dikategorikan ke dalam berbagai "versi" Web:
<Tabs queryString="web-version">
  <TabItem value="web1" label="Web 1.0">
    Web 1.0 adalah "versi" pertama dari Web, yaitu generasi pertama situs web yang muncul ketika Internet mulai tersedia untuk masyarakat luas sekitar tahun 1990.  

>     Web 1.0 sering disebut sebagai web "statis" atau "hanya baca", yang sebagian besar terdiri dari halaman HTML statis sederhana dengan interaktivitas yang terbatas.

Reviewer: ```suggestion
    Web 1.0 sering disebut sebagai web "statis" atau "read-only", yang sebagian besar terdiri dari halaman HTML statis sederhana dengan interaktivitas yang terbatas.
```

    Namun, beberapa kemampuan Web 2.0 sudah ada pada era Web 1.0 tetapi diimplementasikan dengan cara yang berbeda.  

>     Sebagai contoh, situs Web 1.0 mungkin memiliki halaman buku tamu untuk komentar pengunjung, daripada bagian komentar di akhir setiap halaman (yang khas pada Web 2.0).

Reviewer: ```suggestion
    Sebagai contoh, situs Web 1.0 mungkin memiliki halaman buku tamu untuk komentar pengunjung, dari pada bagian komentar di akhir setiap halaman (seperti pada Web 2.0).
```

Author: The correct writing is “daripada”, not “dari pada”. This writing is in accordance with the Indonesian language rules set by the Big Indonesian Dictionary (KBBI)

  </TabItem>
  <TabItem value="web2" label="Web 2.0">
    Pada akhir 1990-an, transisi ke [Web 2.0](https://en.wikipedia.org/wiki/Web_2.0) dimulai, ketika beberapa situs Web 1.0 memperkenalkan HTML dinamis dengan fitur sosial, memungkinkan pengguna berinteraksi dengan halaman web dengan cara yang jauh lebih lancar.
    :::info

>     Istilah "Web 2.0" diciptakan oleh Darcy DiNucci, seorang konsultan arsitektur informasi, dalam artikelnya yang terbit pada Januari 1999 berjudul [Fragmented Future](http://darcyd.com/fragmented_future.pdf).    :::

Reviewer: ```suggestion
    Istilah "Web 2.0" diciptakan oleh Darcy DiNucci, seorang konsultan information architecture, dalam artikelnya yang terbit pada Januari 1999 berjudul [Fragmented Future](http://darcyd.com/fragmented_future.pdf).
    :::
```

    Web 2.0 adalah era di mana media sosial, [aplikasi web](https://en.wikipedia.org/wiki/Web_application), dan aplikasi mobile mulai berkembang pesat.  
    Karena fitur sosial baru dan pengalaman pengguna yang lebih baik yang ditawarkan oleh aplikasi Web 2.0, jumlah pengguna yang berinteraksi dengan web meningkat secara signifikan, yang mengarah pada terciptanya banyak layanan dan aplikasi web baru.  
    Beberapa contoh aplikasi Web 2.0 yang paling populer adalah Facebook, X, Amazon, dan YouTube.

    Banyak dari aplikasi dan layanan ini menjadi pilar penting bagi internet, kehidupan masyarakat, dan bahkan bagi masyarakat secara keseluruhan.  
    Namun, semakin relevan aplikasi-aplikasi ini, semakin jelas pula keterbatasannya, karena pengguna harus mempercayakan data dan privasi mereka kepada perusahaan di baliknya.

>     ##### Keterbatasan Web 2.0

Reviewer: ```suggestion

    ##### Keterbatasan Web 2.0
```

    Meskipun Web 2.0 merevolusi cara orang menggunakan web secara positif, beberapa karakteristiknya menyebabkan beberapa kelemahan bagi pengguna yang berinteraksi dengan aplikasi Web2. Misalnya:

    - Sensor
    - Kurangnya transparansi
    - Privasi terbatas 
    - Interoperabilitas terbatas
  </TabItem>
  <TabItem value="web3" label="Web3" default>

>    Web3 (juga dikenal sebagai Web 3.0) adalah langkah berikutnya dalam evolusi Web, yang bertujuan untuk mengatasi [keterbatasan Web 2.0](web3-app-development?web-version=web2#web-20-limitations).  

Reviewer: ```suggestion
   Web3 (juga dikenal sebagai Web 3.0) adalah langkah berikutnya dalam evolusi Web, yang bertujuan untuk mengatasi [keterbatasan Web 2.0](web3-app-development.mdx?web-version=web2#web2-limitations).  
```

    :::

    Tahun 2015 menandai awal era Web3 ketika Ethereum diluncurkan sebagai blockchain pertama di dunia yang dapat diprogram.  
    Aplikasi Web3 adalah aplikasi terdesentralisasi (juga dikenal sebagai "dApps") yang berjalan di blockchain, memungkinkan pengguna untuk berinteraksi satu sama lain dan dengan blockchain secara trustless.  
    Hal ini memberi pengguna kendali atas data dan privasi mereka, menghilangkan kebutuhan untuk mempercayai otoritas pusat.  

    Aplikasi Web3 dibangun di atas teknologi blockchain, yang memungkinkan pengguna memiliki kepemilikan dan kontrol lebih besar atas data mereka.  
    Ini menghadirkan fitur baru dan memungkinkan layanan yang tidak dapat dilakukan di dunia Web2.  
    Seringkali, aplikasi Web3 menggabungkan teknologi mutakhir lainnya seperti AI untuk memberikan pengalaman pengguna yang unik dan inovatif.

    Jenis baru aplikasi Web3 meliputi:
    
    - DeFi (Decentralized Finance), sebagai contoh:
      - DEXs (Decentralized Exchanges)
      - GameFi (Gaming Finance)
      - Platform staking dan yield farming
      - Platform crowdfunding dan investasi terdesentralisasi

>       - Peminjaman dan peminjaman terdesentralisasi

Reviewer: @fujianggara10 Please kindly proofread each sentence line by line again after you use any translation tool out there.

```suggestion
      - Lending dan borrowing terdesentralisasi
```

Aplikasi mobile dapat memanfaatkan fitur bawaan perangkat genggam dan biasanya lebih cepat serta lebih ramah pengguna dibandingkan dengan situs web.  
Hal ini dapat memberikan pengalaman yang lebih imersif bagi pengguna.  
Terakhir, aplikasi mobile memungkinkan pengguna mengakses aplikasi Anda di mana saja, yang bisa menjadi keuntungan besar untuk berbagai kasus penggunaan.

##### Panduan
- [Bangun Aplikasi Web3 Mobile dengan React Native di Lisk menggunakan thirdweb](https://blog.thirdweb.com/web3-mobile-apps/build-web3-mobile-apps-with-reactive-native-on-lisk/)
- [Video tutorial: Cara Membangun Aplikasi Web3 Mobile](https://www.youtube.com/watch?v=PRAsc_ocVtk) *by thirdweb*
- [Membangun dApp mobile dengan Flutter](https://medium.com/dash-community/building-a-mobile-dapp-with-flutter-be945c80315a)
</TabItem>
<TabItem value="web" label="Aplikasi Web3 Berbasis Web">
Keuntungan utama dari aplikasi web adalah sifatnya yang independen terhadap platform dan dapat diakses dari perangkat apa pun dengan browser web.  
Selain itu, pengguna tidak perlu menginstalnya, karena aplikasi dapat digunakan langsung di browser.

> ##### Panduang

Reviewer: @fujianggara10 Again, need to proofread here..

```suggestion
##### Panduan
```

- [Cara Membangun dApp Minting NFT](https://blog.thirdweb.com/guides/build-nft-minting-dapp-with-thirdweb-guide/) *by thirdweb*
</TabItem>
<TabItem value="pwa" label="Web3 PWA">
Progressive Web App (PWA) memungkinkan pengguna untuk menginstal aplikasi web di perangkat mereka.  
PWA terlihat seperti aplikasi native di desktop atau perangkat mobile dan menawarkan fitur seperti dukungan offline serta notifikasi push.  
Dengan cara ini, PWA memungkinkan pengembangan aplikasi yang dapat digunakan melalui browser, aplikasi mobile, atau desktop secara bersamaan.  
Oleh karena itu, PWA merupakan pilihan yang sangat baik untuk menjangkau pengguna di berbagai platform secara bersamaan.

Selain itu, PWA memungkinkan pengguna untuk menginstal aplikasi langsung melalui web app tanpa perlu mengunjungi app store.  
Hal ini memungkinkan Anda untuk mendistribusikan aplikasi langsung ke pengguna tanpa harus mematuhi batasan yang diberlakukan oleh app store terhadap aplikasi berbasis blockchain.  

> Biaya pembelian dalam aplikasi juga dapat dihindari karena pengguna tidak lagi perlu melalui app store.  

Reviewer: ```suggestion
Biaya pembelian dalam aplikasi juga dapat dihindari karena pengguna tidak lagi perlu membayar melalui app store.  
```

> [^1]: Hanya didukung di browser yang kompatibel dengan PWA (sebagian besar browser berbasis Chromium seharusnya berfungsi).

Reviewer: ```suggestion
[^1]: Hanya didukung di browser yang kompatibel dengan PWA (sebagian besar browser berbasis Chromium seharusnya mendukungnya).
```

Untuk membawa pengguna Web2 ke Web3, aplikasi perlu menyederhanakan interaksi pengguna agar terasa alami dan intuitif, sekaligus tetap memberikan manfaat Web3 dengan cara yang aman.  

### **Account Abstraction dan Smart Accounts**  

Account abstraction adalah konsep yang menambahkan lapisan tambahan di atas blockchain, yang mendelegasikan sebagian besar kompleksitas interaksi blockchain dari pengguna ke smart contracts. Ini secara signifikan meningkatkan pengalaman pengguna aplikasi Web3 karena memungkinkan interaksi yang lebih mudah dan aman.  

Kadang-kadang disebut sebagai **"smart accounts"** atau **"smart wallets"**, konsep ini memungkinkan sebagian besar logika transaksi dan interaksi pengguna dikelola oleh smart contracts, membuka fitur-fitur berikut:  

- **Signature Abstraction**: Pengguna tidak perlu lagi menandatangani transaksi secara manual dengan private key. Smart contracts menangani logika tanda tangan, mengurangi risiko kesalahan pengguna dan serangan phishing.  
- **Keamanan Lebih Tinggi**: Implementasi dompet multi-signature, mekanisme pemulihan sosial, dan kontrol akses kustom dalam smart contracts dapat mengurangi risiko pencurian dan kegagalan tunggal.  

> - **Transaksi yang Dapat Disesuaikan**: Smart contracts dapat diprogram untuk menetapkan batas pengeluaran, transaksi dengan waktu terkunci, dan kontrol khusus untuk dApps, memungkinkan otomatisasi transaksi.  

Reviewer: ```suggestion
- **Transaksi yang Dapat Disesuaikan**: Smart contracts dapat diprogram untuk menetapkan batas pengeluaran, transaksi dengan batasan waktu, dan kontrol khusus untuk dApps, memungkinkan otomatisasi transaksi.  
```

- **Fee Abstraction**: Biaya transaksi dapat dikelola secara dinamis melalui smart contracts, memungkinkan fitur seperti token gas kustom dan transaksi tanpa gas untuk kasus penggunaan tertentu.  

Dari sisi teknis, account abstraction diimplementasikan oleh [standar ERC-4337](https://eips.ethereum.org/EIPS/eip-4337), yang mendefinisikan serangkaian antarmuka dan fungsi yang dapat digunakan oleh smart contracts untuk berinteraksi dengan blockchain atas nama pengguna.

Ini terdiri dari komponen fundamental berikut:

- **UserOperation**: Struktur yang mendeskripsikan transaksi yang akan dikirim atas nama pengguna.  
  Untuk menghindari kebingungan, ini tidak disebut sebagai "transaksi".  
  Seperti transaksi, ia berisi `sender`, `to`, `calldata`, `maxFeePerGas`, `maxPriorityFee`, `signature`, `nonce`.  
  Tidak seperti transaksi, penggunaan bidang `signature` tidak ditentukan oleh protokol tetapi oleh setiap implementasi akun.  
  Selain itu, ia berisi beberapa bidang lain yang dijelaskan secara rinci dalam ERC-4337.  

- **Account Contract**: Smart account pengguna yang mengirimkan UserOperation.  

> - **EntryPoint**: Kontrak tunggal yang memvalidasi dan mengeksekusi kumpulan UserOperations.  

Reviewer: ```suggestion
- **EntryPoint**: Contract tunggal yang memvalidasi dan mengeksekusi kumpulan UserOperations.  
```

>   Kontrak ini mengelola logika verifikasi dan eksekusi untuk UserOperations, memastikan mereka diproses dengan benar.  

Reviewer: ```suggestion
  Contract ini mengelola logika verifikasi dan eksekusi untuk UserOperations, memastikan mereka diproses dengan benar.  
```

> - **Bundler**: Bundlers adalah kontrak yang mengemas UserOperations dari mempool dan mengirimkannya ke EntryPoint.  

Reviewer: ```suggestion
- **Bundler**: Bundlers adalah contract yang mengemas UserOperations dari mempool dan mengirimkannya ke EntryPoint.  
```

  Bundling adalah proses di mana sebuah node/bundler mengumpulkan beberapa UserOperations dan membuat satu transaksi untuk dikirimkan ke blockchain.  

> - **Account Factory**: Saat menggunakan dompet untuk pertama kali, kontrak ini menentukan pembuatan smart contract wallet.  

Reviewer: ```suggestion
- **Account Factory**: Saat menggunakan dompet untuk pertama kali, contract ini menentukan pembuatan smart contract wallet.  
```

> - **Paymaster** (opsional): Kontrak pembantu yang bersedia membayar transaksi, bukan pengirimnya sendiri.  

Reviewer: ```suggestion
- **Paymaster** (opsional): Contract pembantu yang bersedia membayar transaksi, bukan pengirimnya sendiri.  
```

  Paymasters dapat menanggung biaya gas untuk pengguna, mensponsori transaksi tertentu, atau menerapkan model biaya khusus, membuka peluang untuk berbagai kasus penggunaan baru.

##### Bacaan lebih lanjut
- [Abstraksi Akun: Gambaran Lengkap](https://blaize.tech/blog/account-abstraction-guide/) *by Blaize*
- [Apa itu Abstraksi Akun (ERC-4337)?](https://www.alchemy.com/overviews/what-is-account-abstraction) *by Alchemy*
- [Apa itu Abstraksi Akun? Penjelasan ERC-4337](https://blog.thirdweb.com/account-abstraction-erc4337/) *by thirdweb*
- [Bagaimana cara membayar biaya gas dengan LSK](https://blog.thirdweb.com/changelog/ts-tokenpaymasters-pay-for-gas-with-lisk-lsk-base-usdc-or-celo-cusd/) *by thirdweb*
- [Bagaimana cara membuat aplikasi Web3 mobile dengan transaksi tanpa gas di Lisk](https://blog.thirdweb.com/web3-mobile-apps/build-web3-mobile-apps-with-reactive-native-on-lisk/) *by thirdweb*

{/* TODO: Perluas dengan topik tambahan yang relevan bagi pengembang Web3
### Praktik Terbaik & Pertimbangan lainnya

> - Kontrak proxy: 

Reviewer: ```suggestion
- Contract proxy: 
```

Smart contract tidak dapat diubah, yang berarti bahwa setelah digunakan, smart contract tidak dapat diubah.

> Akan tetapi, untuk memperbaiki bug atau menambahkan fitur baru pada smart contract yang sudah digunakan, kontrak proksi dapat digunakan untuk meng-upgrade smart contract tanpa mengubah alamat kontrak.

Reviewer: ```suggestion
Akan tetapi, untuk memperbaiki bug atau menambahkan fitur baru pada smart contract yang sudah digunakan, contract proxy dapat digunakan untuk meng-upgrade smart contract tanpa mengubah alamat contract.
```

- Optimalisasi Gas:
- manajemen kunci yang aman
- Program bug bounty
- menangani data pengguna */}

> ### Pengujian dan penerapan

Reviewer: ```suggestion
### Pengujian dan deployment
```

> Dalam hal pengujian dan penerapan aplikasi Web3, ada beberapa aspek tambahan yang perlu dipertimbangkan dibandingkan dengan aplikasi Web2.

Reviewer: ```suggestion
Dalam hal pengujian dan deployment aplikasi Web3, ada beberapa aspek tambahan yang perlu dipertimbangkan dibandingkan dengan aplikasi Web2.
```

Reviewer: @fujianggara10 Please also note that @TalhaMaliktz and I have decided that every variation of the word 'deploy' will not be translated to Bahasa Indonesia (luncur, terap, etc). Please retain every occurence of the word 'deploy' and its variations in all the docs translations.

> Kerentanan intrinsik dari smart contract, kekekalan data blockchain, dan interaksi dengan jaringan yang terdesentralisasi, membutuhkan pengujian dan langkah-langkah keamanan yang spesifik.

Reviewer: ```suggestion
Vulnerability intrinsik dari smart contract, immutability data blockchain, dan interaksi dengan jaringan yang terdesentralisasi, membutuhkan pengujian dan langkah-langkah keamanan yang spesifik.
```

Untungnya, banyak alat yang sudah tersedia untuk membantu para pengembang memastikan keamanan dan keandalan aplikasi Web3 mereka.
Daftar di bawah ini menyediakan pilihan alat dan layanan populer yang digunakan untuk menguji aplikasi Web3:

> #### Rangkaian pengujian kontrak pintar

Reviewer: ```suggestion
#### Rangkaian pengujian smart contract
```

> Sering kali, kerangka kerja pengembangan smart contract dilengkapi dengan alat pengujian bawaan yang menyederhanakan untuk membuat pengujian komprehensif untuk smart contract.

Reviewer: ```suggestion
Sering kali, framework pengembangan smart contract dilengkapi dengan alat pengujian bawaan yang menyederhanakan untuk membuat pengujian komprehensif untuk smart contract.
```

Ini termasuk utilitas untuk menyederhanakan penulisan pengujian, atau kemampuan untuk membuat node lokal untuk menyebarkan dan menguji smart contract dalam jaringan pengembangan lokal, yaitu Devnet.

> Pilihan lingkungan pengujian populer untuk smart contract dapat ditemukan di bawah ini:

Reviewer: ```suggestion
Pilihan environment pengujian populer untuk smart contract dapat ditemukan di bawah ini:
```

@fujianggara10 Let's not translate environment as well, as it is a common technical word and weird if it is translated.

> - [Foundry](https://book.getfoundry.sh/): Foundry adalah sebuah rantai alat pengembangan kontrak pintar.

Reviewer: ```suggestion
- [Foundry](https://book.getfoundry.sh/): Foundry adalah sebuah rantai alat pengembangan smart contract.
```

> Foundry menyediakan [kerangka kerja pengujian] yang kuat (https://book.getfoundry.sh/forge/tests) yang memungkinkan pengembang untuk membuat rangkaian pengujian yang komprehensif untuk proyek mereka menggunakan Solidity.

Reviewer: ```suggestion
Foundry menyediakan [framework pengujian] yang kuat (https://book.getfoundry.sh/forge/tests) yang memungkinkan pengembang untuk membuat rangkaian pengujian yang komprehensif untuk proyek mereka menggunakan Solidity.
```

> [Perpustakaan Standar Forge](https://book.getfoundry.sh/reference/forge-std/) menyediakan seperangkat utilitas untuk menyederhanakan pengujian smart contract.

Reviewer: ```suggestion
[Library Standar Forge](https://book.getfoundry.sh/reference/forge-std/) menyediakan seperangkat utilitas untuk menyederhanakan pengujian smart contract.
```

> Ini juga menyediakan simpul testnet lokal untuk menerapkan dan menguji kontrak pintar, lihat [Anvil](https://book.getfoundry.sh/anvil/).

Reviewer: ```suggestion
Ini juga menyediakan simpul testnet lokal untuk mendeploy dan menguji smart contract, lihat [Anvil](https://book.getfoundry.sh/anvil/).
```

> - [Hardhat](https://hardhat.org): Hardhat adalah sebuah lingkungan pengembangan yang memfasilitasi pengembangan smart contract.

Reviewer: ```suggestion
- [Hardhat](https://hardhat.org): Hardhat adalah sebuah environment pengembangan yang memfasilitasi pengembangan smart contract.
```

Hardhat dilengkapi dengan [Hardhat Network](https://hardhat.org/hardhat-network/docs/overview), sebuah node jaringan Ethereum lokal yang didesain untuk pengembangan.

> Hardhat memungkinkan Anda untuk menggunakan kontrak Anda, [menjalankan pengujian Anda](https://hardhat.org/tutorial/testing-contracts), dan men-debug kode Anda, semuanya dalam batasan mesin lokal Anda.

Reviewer: ```suggestion
Hardhat memungkinkan Anda untuk menggunakan contract Anda, [menjalankan pengujian Anda](https://hardhat.org/tutorial/testing-contracts), dan men-debug kode Anda, semuanya dalam local machine Anda.
```

Pengujian ditulis dalam JavaScript, menggunakan [ethers](https://ethers.org/).

> - [Remix](https://remix.ethereum.org/): Remix adalah sebuah IDE yang membantu Anda menulis kontrak Solidity langsung dari browser.

Reviewer: ```suggestion
- [Remix](https://remix.ethereum.org/): Remix adalah sebuah IDE yang membantu Anda menulis contract Solidity langsung dari browser.
```

Reviewer: @fujianggara10 @TalhaMaliktz and I have decided that any wording of contract / smart contract will not be translated to Bahasa Indonesia. Please keep this in mind for future translations~

Remix menyediakan fitur-fitur yang kuat untuk menguji dan men-debug smart contract, seperti [Debugger](https://remix-ide.readthedocs.io/en/latest/debugger.html) atau [Unit Testing Plugin](https://remix-ide.readthedocs.io/en/latest/unittesting.html).

> - [Waffle](https://getwaffle.io/): Waffle adalah kerangka kerja lain untuk mengembangkan dan menguji smart contract.

Reviewer: ```suggestion
- [Waffle](https://getwaffle.io/): Waffle adalah framework lain untuk mengembangkan dan menguji smart contract.
```

Pengujian ditulis dalam JavaScript / Typescript.

##### Panduan

> - [Menerapkan kontrak pintar dengan Foundry](/building-on-lisk/menerapkan-kontrak-pintar/dengan-Foundry#menguji-kontrak-pintar)

Reviewer: @fujianggara10 This is a link to other page, therefore the path doesn't need to be translated, except for the internal link starting with #. In this case #menguji-smart-contract. Please make sure that the link is pointing to the correct file, and the linked heading also exists.

```suggestion
- [Mendeploy smart contract dengan Foundry](/building-on-lisk/deploying-smart-contract/with-Foundry#menguji-smart-contract)
```

> - [Menerapkan kontrak pintar dengan Hardhat](/building-on-lisk/menerapkan-kontrak-pintar/dengan-Hardhat.md)

Reviewer: ```suggestion
- [Mendeploy smart contract dengan Hardhat](/building-on-lisk/deploying-smart-contract/with-Hardhat.md)
```

> - [Menerapkan kontrak pintar dengan Remix](/building-on-lisk/menerapkan-kontrak-pintar/dengan-Remix.mdx)

Reviewer: ```suggestion
- [Mendeploy smart contract dengan Remix](/building-on-lisk/deploying-smart-contract/with-Remix.mdx)
```

> #### Analisis kode kontrak pintar

Reviewer: ```suggestion
#### Analisis kode smart contract
```

> Untuk memastikan keamanan smart contract, sangat penting untuk menganalisis kode untuk mengetahui adanya kerentanan.

Reviewer: ```suggestion
Untuk memastikan keamanan smart contract, sangat penting untuk menganalisis kode untuk mengetahui adanya vulnerability.
```

Reviewer: @fujianggara10 @TalhaMaliktz Let's also retain vulnerability word and no need to translate it to Bahasa Indonesia. Please note that in Indonesian language, plural and singular words are the same, so there will be occurence where the English is vulnerabilities but the translated word will become vulnerability.

> Ada beberapa alat yang tersedia yang membantu pengembang untuk menganalisis smart contract mereka untuk masalah keamanan umum dan kerentanan, misalnya:

Reviewer: ```suggestion
Ada beberapa alat yang tersedia yang membantu pengembang untuk menganalisis smart contract mereka untuk masalah keamanan umum dan vulnerability, misalnya:
```

> - [Slither](https://github.com/crytic/slither): Slither adalah kerangka kerja analisis statis Solidity yang mendeteksi kerentanan umum dalam smart contract.

Reviewer: ```suggestion
- [Slither](https://github.com/crytic/slither): Slither adalah framework analisis statis Solidity yang mendeteksi vulnerability umum dalam smart contract.
```

- [Etheno](https://github.com/crytic/etheno): Etheno adalah sebuah alat yang membantu pengembang untuk menulis smart contract yang aman dengan menyediakan seperangkat aturan yang dapat digunakan untuk memeriksa keamanan smart contract.
- [Mythril](https://github.com/Consensys/mythril): Mythril adalah alat analisis keamanan untuk smart contract Ethereum.

#### Testnets
Testnet adalah blockchain publik yang digunakan untuk tujuan pengujian.

> Testnet sering kali memiliki fitur yang sama dengan mainnet, tetapi token pada testnet tidak memiliki nilai nyata - ini memungkinkan pengembang untuk menguji smart contract dan aplikasi Web3 mereka di lingkungan dunia nyata tanpa mempertaruhkan aset nyata.

Reviewer: ```suggestion
Testnet seringkali memiliki fitur yang sama dengan mainnet, tetapi token pada testnet tidak memiliki nilai nyata - ini memungkinkan pengembang untuk menguji smart contract dan aplikasi Web3 mereka di dunia nyata tanpa mempertaruhkan aset nyata.
```

> Sebelum menerapkan smart contract ke mainnet, sangat disarankan untuk mengujinya di testnet untuk memastikan bahwa smart contract bekerja sesuai dengan yang diharapkan.

Reviewer: ```suggestion
Sebelum deploy smart contract ke mainnet, sangat disarankan untuk mengujinya di testnet untuk memastikan bahwa smart contract bekerja sesuai dengan yang diharapkan.
```

##### Testnet publik
Testnet berikut ini tersedia untuk menguji smart contract dan aplikasi Web3 di Lisk:
- [Lisk Sepolia](/about-lisk/network-info.md#lisk-sepolia-testnet)

:::tip

> Untuk mendapatkan testnet ETH dan/atau LSK gratis, lihat [faucet yang tersedia] /lisk-tools/faucets.md).

Reviewer: ```suggestion
Untuk mendapatkan testnet ETH dan/atau LSK gratis, lihat [faucet yang tersedia](/lisk-tools/faucets.md).
```

:::

##### Testnet virtual
[Tenderly](https://tenderly.co/) baru-baru ini memperkenalkan kemungkinan untuk menjalankan [Virtual Testnets](https://blog.tenderly.co/how-virtual-testnets-replace-public-testnets/).

> Virtual Testnets adalah cara baru untuk menguji aplikasi Web3 Anda, membawa data onchain dari mainnet dan fleksibilitas lingkungan lokal bersama-sama.

Reviewer: ```suggestion
Virtual Testnets adalah cara baru untuk menguji aplikasi Web3 Anda, membawa data onchain dari mainnet dan fleksibilitas environment lokal bersama-sama.
```

Beberapa fitur menarik dari Virtual Testnets adalah:

- **Bangun dapps Anda dengan data mainnet yang nyata dan terkini** dan uji performa aktualnya dalam kondisi yang realistis. 
- **Dapatkan faucet tanpa batas** dan berhenti mengemis untuk uji coba ETH.
Cetak token asli dan ERC-20 dalam jumlah berapa pun yang Anda butuhkan, meniru akun, dan menguji skenario kompleks tanpa gangguan.
- Sesuaikan sepenuhnya infrastruktur pengembangan Anda agar sesuai dengan alur kerja unik Anda.

> Virtual Testnet mudah diatur dan dikonfigurasi, memberi Anda kontrol penuh dan fleksibilitas atas lingkungan Anda.

Reviewer: ```suggestion
Virtual Testnet mudah diatur dan dikonfigurasi, memberi Anda kontrol penuh dan fleksibilitas atas environment Anda.
```

> Anda bahkan dapat memanipulasi status rantai bercabang Anda dengan nilai khusus, ID rantai khusus, dan banyak lagi, memastikannya sesuai dengan kebutuhan unik Anda.

Reviewer: ```suggestion
Anda bahkan dapat memanipulasi status forked chain Anda dengan nilai custom, chain ID custom, dan banyak lagi, memastikannya sesuai dengan kebutuhan unik Anda.
```

@fujianggara10 we should preserve the word chain, no need to translate it to rantai as this is a common technical web3 term.

#### Pengujian E2E
Pengujian end-to-end (E2E) adalah jenis pengujian perangkat lunak yang menguji alur lengkap aplikasi dari awal hingga akhir.
Hal ini secara khusus mencakup pengujian frontend aplikasi Anda, yang memastikan bahwa antarmuka pengguna berfungsi seperti yang diharapkan.
Untuk menangani hal-hal seperti dompet, dan hal-hal khusus Web3 lainnya, Anda mungkin ingin menggunakan alat pengujian Web3 E2E khusus.

> [Synpress](https://synpress.io/) adalah perpustakaan pengujian E2E untuk dapps Web3. 

Reviewer: ```suggestion
[Synpress](https://synpress.io/) adalah library pengujian E2E untuk dapps Web3. 
```

@fujianggara10 always use the word library instead of perpustakaan, as perpustakaan will refer to an actual physical library in Indonesian language.

> Synpress dapat digunakan dengan dua kerangka kerja pengujian E2E terkemuka di luar sana, yaitu, [Playwright](https://playwright.dev/) dan [Cypress](https://www.cypress.io/), dengan dukungan untuk semua fitur mereka.

Reviewer: ```suggestion
Synpress dapat digunakan dengan dua framework pengujian E2E terkemuka di luar sana, yaitu, [Playwright](https://playwright.dev/) dan [Cypress](https://www.cypress.io/), dengan dukungan untuk semua fitur mereka.
```

#### Pengujian kinerja
Pengujian performa sangat penting untuk memastikan bahwa aplikasi Web3 Anda dapat menangani beban yang diharapkan dari pengguna dan interaksi mereka.

> Ukur dan tingkatkan performa, throughput, dan ketahanan aplikasi Anda di bawah tingkat lalu lintas dan transaksi yang berbeda.

Reviewer: ```suggestion
Ukur dan tingkatkan performa, throughput, dan ketahanan aplikasi Anda di bawah level traffic dan transaksi yang berbeda.
```

Beberapa contoh alat pengujian kinerja adalah:

- [k6](https://k6.io/)
- [Artillery](https://www.artillery.io/)

{/* #### Pengujian Otomatis dan Integrasi Berkelanjutan (CI) */}

#### Pemantauan
Pemantauan penting untuk memastikan bahwa aplikasi Web3 Anda berjalan dengan lancar dan untuk mendeteksi dan memperbaiki masalah sebelum mempengaruhi pengguna Anda.

Beberapa alat pemantauan yang populer adalah:

- [Prometheus](https://prometheus.io/) yang dikombinasikan dengan [Grafana](https://grafana.com/)
- [Elastic Stack (ELK)](https://www.elastic.co/elastic-stack)

### Pengembangan aplikasi Web2 vs Web3
#### Bahasa
Mengenai pengembangan aplikasi frontend, tidak banyak perbedaan antara aplikasi Web2 dan Web3 dalam hal bahasa pemrograman.
Semua bahasa yang biasanya digunakan untuk pengembangan aplikasi Web2 juga dapat digunakan untuk pengembangan aplikasi Web3.
Pada akhirnya, aplikasi Web3 bahkan dapat terlihat dan terasa persis seperti aplikasi Web2 bagi pengguna, karena frontend mereka dibangun dengan teknologi yang sama.

> Namun, bagian belakang aplikasi Web3 sangat berbeda, karena berinteraksi dengan blockchain melalui kontrak pintar.

Reviewer: ```suggestion
Namun, bagian belakang aplikasi Web3 sangat berbeda, karena berinteraksi dengan blockchain melalui smart contract.
```

> Kontrak pintar ditulis dalam [Solidity](https://soliditylang.org/), oleh karena itu pengembang Web3 perlu mengetahui Solidity untuk membangun kontrak pintar yang mereka perlukan untuk aplikasi mereka.

Reviewer: ```suggestion
Smart contract ditulis dalam [Solidity](https://soliditylang.org/), oleh karena itu pengembang Web3 perlu mengetahui Solidity untuk membangun smart contract yang mereka perlukan untuk aplikasi mereka.
```

Untungnya, untuk banyak kasus penggunaan, sudah tersedia smart contract yang telah diaudit (lihat [OpenZeppelin](https://www.openzeppelin.com/solidity-contracts) atau [thirdweb](https://thirdweb.com/explore)) yang dapat digunakan, sehingga pengembang tidak perlu menulis semua smart contract dari awal.

> #### Tumpukan teknologi

Reviewer: ```suggestion
#### Tech stack
```

Ada juga SDK untuk pengembangan aplikasi Web3, seperti [thirdweb](https://thirdweb.com/) atau [Alchemy](https://www.alchemy.com/), yang menghilangkan sebagian besar kerumitan pengembangan blockchain dan memberikan pengalaman pengembang yang lancar.
thirdweb mendukung [Lisk](https://thirdweb.com/lisk) dan banyak blockchain yang kompatibel dengan EVM lainnya.

#### Penyimpanan data
Dalam aplikasi Web2, data biasanya disimpan dalam basis data terpusat.
Pada aplikasi Web3, beberapa data disimpan di blockchain, yaitu basis data yang terdesentralisasi.
Bagaimana data disimpan ditentukan di dalam smart contract.
Namun, perlu dicatat bahwa data onchain pada blockchain publik dapat dilihat oleh semua orang dan, oleh karena itu, tidak cocok untuk semua jenis data.
Terutama data pengguna yang sensitif dan pribadi tidak boleh disimpan di blockchain, kecuali jika data tersebut dienkripsi.
Selain itu, menyimpan data dalam jumlah besar di blockchain itu mahal dan lambat.
Oleh karena itu, aplikasi Web3 terkadang menggunakan kombinasi penyimpanan onchain dan penyimpanan offchain.

#### Akses data
Pada aplikasi Web2, data dari database diakses melalui API.

> Pada aplikasi Web3, data blockchain diakses melalui kontrak pintar.

Reviewer: ```suggestion
Pada aplikasi Web3, data blockchain diakses melalui smart contract.
```

> Kontrak ini biasanya mengekspos fungsi publik yang dapat dipanggil oleh frontend aplikasi untuk membaca dan/atau menulis data dari/ke blockchain, mirip dengan API klasik.

Reviewer: ```suggestion
Contract ini biasanya mengekspos fungsi publik yang dapat dipanggil oleh frontend aplikasi untuk membaca dan/atau menulis data dari/ke blockchain, mirip dengan API klasik.
```
